src/lib/y2partitioner/widgets/mkfs_optiondata.rb

```diff
@@ -0,0 +1,254 @@
+require "yast"
+
+module Y2Partitioner
+  module Widgets
+    # a class
+    class MkfsOptiondata
```

Comment:
If you never create an instance of this think (and you doesn't), then this should not be a class. I'm still trying to wrap my head around the whole thing to decide what it should be. :-)

Reply:
It would make more sense from the object orientation POV if every instance of this class would correspond to each entry of the `FOO` array. So you could do something like:
```ruby
options = MkfsOptiondata.options_for(filesystem) # Array of objects, no array of hashes
option.first.default #=> "none"
options.first.default_is_boolean? #=> More readable than that case condition you have down there in other part of the code
...
```

Reply:
All this seems to be designed to, somehow, avoid too much subclassing. For example, two comboboxes should be of the corresponding ComboBox class, even if they access different properties or have an slightly different validation. But the philosophy of the CWM OOP layer is (as weird as it sounds at first) to have a separate class for each widget. So you are expected to actually define something like `BlockSize < ComboBox` and `BytesPerInode < ComboBox`.

If you have a different class for each widget everything will flow better and you shouldn't need that many complex static methods, custom handlers for `help` or hacks like `handle_all_events = true`.

So, making a real class from each one of the entries in that FOO array would be the correct `CWM` approach. IMHO.

Josef is the authority in OOP-CWM though, if you want a second opinion. :wink:

Reply:
> I'm still trying to wrap my head around the whole thing to decide what it should be. :-)

Well, in `widgets/help.rb` you made it a module. :-)

I'm not convinced I should turn the hashes into classes just for the nicer look  ATM the code is small and still easy to follow.

> But the philosophy of the CWM OOP layer is [...] to have a separate class for each widget

To have a unique `#widget_id` seems to be the requirement.

Getting the help button to work was strange, though. The code is copied from the fstab dialog. I haven't seen any other help button in the expert partitioner so far.